src/ray/raylet/task_dependency_manager.cc

```diff
@@ -196,23 +208,67 @@ void TaskDependencyManager::TaskPending(const Task &task) {
   TaskID task_id = task.GetTaskSpecification().TaskId();
 
   // Record that the task is pending execution.
-  pending_tasks_.insert(task_id);
-  // Find any subscribed tasks that are dependent on objects created by the
-  // pending task.
-  auto remote_task_entry = required_tasks_.find(task_id);
-  if (remote_task_entry != required_tasks_.end()) {
-    for (const auto &object_entry : remote_task_entry->second) {
-      // This object created by the pending task will appear locally once the
-      // task completes execution. Cancel any in-progress operations to make
-      // the object local.
-      HandleRemoteDependencyCanceled(object_entry.first);
+  auto inserted =
+      pending_tasks_.emplace(task_id, PendingTask(initial_lease_period_ms_, io_service_));
+  if (inserted.second) {
+    // This is the first time we've heard that this task is pending.  Find any
+    // subscribed tasks that are dependent on objects created by the pending
+    // task.
+    auto remote_task_entry = required_tasks_.find(task_id);
+    if (remote_task_entry != required_tasks_.end()) {
+      for (const auto &object_entry : remote_task_entry->second) {
+        // This object created by the pending task will appear locally once the
+        // task completes execution. Cancel any in-progress operations to make
+        // the object local.
+        HandleRemoteDependencyCanceled(object_entry.first);
+      }
     }
+
+    // Acquire the lease for the task's execution in the global lease table.
+    AcquireTaskLease(task_id);
   }
 }
 
+void TaskDependencyManager::AcquireTaskLease(const TaskID &task_id) {
+  auto it = pending_tasks_.find(task_id);
+  int64_t now_ms = current_time_ms();
+  if (it == pending_tasks_.end()) {
+    return;
+  }
+
+  // Check that we were able to renew the task lease before the previous one
+  // expired.
+  if (now_ms > it->second.expires_at) {
+    RAY_LOG(WARNING) << "Task lease to renew has already expired by "
+                     << (it->second.expires_at - now_ms) << "ms";
+  }
+
+  auto task_lease_data = std::make_shared<TaskLeaseDataT>();
+  task_lease_data->node_manager_id = client_id_.hex();
+  task_lease_data->acquired_at = current_sys_time_ms();
+  task_lease_data->timeout = it->second.lease_period;
+  RAY_CHECK_OK(task_lease_table_.Add(DriverID::nil(), task_id, task_lease_data, nullptr));
```

Comment:
Actually, it looks like the task lease table is indexed by task ID, which seems good. This is the only place where we add to the table, right?

Reply:
Right! That was a typo.